### PR TITLE
 Add map_index to XCom model and interface

### PR DESCRIPTION
**Pull request by uranusjr**
This adds an additional (optional) map_index argument to XCom's get/set/clear interface so mapped task instances can push to the correct entries, and have them pulled correctly by a downstream.

To make the XCom interface easier to use for common scenarios, a convenience method get_value is added to take a TaskInstanceKey that automatically performs argument unpacking and call get_one underneath. This is not done as a get_one overload to simplify the implementation and typing.

### CODE REVIEW

**Comment by uranusjr** [comment]
Wow this conflicts very fast.

File: airflow/migrations/versions/c306b5b5ae4a_switch_xcom_table_to_use_run_id.py [hunk 1]
```diff
@@ -132,6 +134,10 @@ def downgrade():
     op.create_table("__airflow_tmp_xcom", *_get_old_xcom_columns())
 
     xcom = Table("xcom", metadata, *_get_new_xcom_columns())
+
+    # Remoe XCom entries from mapped tis.
+    delete(xcom).where(xcom.c.map_index != -1)
```
Comment (ashb) [inline airflow/migrations/versions/c306b5b5ae4a_switch_xcom_table_to_use_run_id.py]: This doesn't go via alembic's op which means it doesn't get captured in to `--sql` mode.

Comment (ashb) [reply airflow/migrations/versions/c306b5b5ae4a_switch_xcom_table_to_use_run_id.py]:
```suggestion
    op.execute(xcom.delete().where(xcom.c.map_index != -1))
```

File: airflow/models/xcom.py [hunk 2]
```diff
@@ -87,7 +88,7 @@ class BaseXCom(Base, LoggingMixin):
         # but it goes over MySQL's index length limit. So we instead create indexes
         # separately, and enforce uniqueness with DagRun.id instead.
         Index("idx_xcom_key", key),
-        Index("idx_xcom_ti_id", dag_id, task_id, run_id),
+        Index("idx_xcom_ti_id", dag_id, task_id, run_id, map_index),
```
Comment (ashb) [inline airflow/models/xcom.py:91]: The column order here doesn't match the migrations -- this matters from one of our DBs (sqlite? mssql?)

Comment (ashb) [reply airflow/models/xcom.py:91]: Shouldn't we have an FK to TaskInstance?

File: airflow/models/xcom.py [hunk 3]
```diff
@@ -203,27 +209,25 @@ def set(
...
+        match the criteria, an arbitrary one is returned.
 
-        A deprecated form of this function accepts ``execution_date`` instead of
-        ``run_id``. The two arguments are mutually exclusive.
-
-        :param ti_key: The TaskInstanceKey to look up the XCom for
+        :param ti_key: The TaskInstanceKey to look up the XCom for.
         :param key: A key for the XCom. If provided, only XCom with matching
             keys will be returned. Pass *None* (default) to remove the filter.
         :param include_prior_dates: If *False* (default), only XCom from the
```
Comment (ashb) [inline airflow/models/xcom.py]: This method doesn't accept include_prior_dates.

File: airflow/providers/amazon/aws/operators/emr.py [hunk 4]
```diff
@@ -245,7 +245,7 @@ def get_link(
         :return: url link
         """
         if ti_key:
-            flow_id = XCom.get_one(key="return_value", ti_key=ti_key)
+            flow_id = XCom.get_value(key="return_value", ti_key=ti_key)
```
Comment (ashb) [inline airflow/providers/amazon/aws/operators/emr.py:248]: We can't use this new method :/ The providers need to continue to work with Airflow 2.1 :/

Comment (ashb) [reply airflow/providers/amazon/aws/operators/emr.py:248]: Gah, no. This is the compat shim that _I_ wrote. Ingore me, this is fine.

**Review by ashb — COMMENTED** [review]
LGTM otherise.

File: airflow/migrations/versions/c306b5b5ae4a_switch_xcom_table_to_use_run_id.py (hunk 1 above)
Comment (ashb) [reply airflow/migrations/versions/c306b5b5ae4a_switch_xcom_table_to_use_run_id.py]: `airflow db downgrade -s --revision a3bcd0914482` to get the SQL.

Comment (uranusjr) [reply airflow/migrations/versions/c306b5b5ae4a_switch_xcom_table_to_use_run_id.py]: Actually the way I wrote it the query won’t get executed at all 🤪

Comment (ashb) [reply airflow/migrations/versions/c306b5b5ae4a_switch_xcom_table_to_use_run_id.py]: :lolsob:

File: airflow/models/xcom.py (hunk 2 above)
Comment (uranusjr) [reply airflow/models/xcom.py:91]: There wasn’t a foreign key before. I think previously (when XCom was based on execution date) it was possible to push an XCom to a future date, so a foreign key did not make sense back then; now it’s based on run ID perhaps it makes sense to have a ti-xcom relation, but that should be a separate discussion regardless.

Comment (ashb) [reply airflow/models/xcom.py:91]: > it was possible to push an XCom to a future date

Already done https://lists.apache.org/thread/gofj3g6m6vvksy6n0cmgq1qxd309bbbl (I don't think it ever _actually_ worked.)

**Comment by dstandish** [comment]
heads up @uranusjr that migration has been renamed to `0102_c306b5b5ae4a_switch_xcom_table_to_use_run_id.py` in main

**Review by ashb — APPROVED** [review]

**Comment by potiuk** [comment]
Hello @uranusjr  @ashb - while preparaing the "Throve classifier" release for Providers i noticed that backwards-incompatible change sneaked in for a number of providers here.

The `get_value` method of XCom is only added in 2.3.0, yet a number of providers have been switched to use it in some places in this change. I think we we should implement a backwards-compatibility here and add protection that this new XCom method iis not used accidentally by the providers.

**Comment by uranusjr** [comment]
Those `get_value()` usages should only happen in `if ti_key is not None` blocks. The `ti_key` argument is only passed by Airflow 2.3+, so this check acts as feature detection to maintain backward compatibility. Please do tell me if you spot any usages that are not in such blocks.

**Comment by potiuk** [comment]
> Those `get_value()` usages should only happen in `if ti_key is not None` blocks. The `ti_key` argument is only passed by Airflow 2.3+, so this check acts as feature detection to maintain backward compatibility. Please do tell me if you spot any usages that are not in such blocks.

Ah ok. Just thinking how to automate the check in this case so that it is not used accidentally. Let me take a look maybe I will come up with something. I do not want to spot it myself, our backwards compatibility checks should do it automatically for us I think. Same way as we check if "imports" are working.

**Comment by potiuk** [comment]
Added PR for that @uranusjr  https://github.com/apache/airflow/pull/22244

**Comment by potiuk** [comment]
(BTW. there were no mis-uses, but I standardized the use of `if not None`

**Pull request merged**